components/http_client.rst

```diff
@@ -0,0 +1,355 @@
+---------------
+
+The client created with the ``HttpClient`` class provides a single ``request()``
+method to perform all HTTP requests, whereas the client created with the
+``ApiClient`` class provides a method for each HTTP verb::
+
+    $response = $httpClient->request('GET', 'https://');
+    $response = $httpClient->request('POST', 'https://');
+    $response = $httpClient->request('PUT', 'https://');
+    // ...
+
+    $response = $apiClient->get('https://');
+    $response = $apiClient->post('https://');
+    $response = $apiClient->put('https://');
+    // ...
+
+Query String Parameters
+~~~~~~~~~~~~~~~~~~~~~~~
+
+You can either append them manually to the requested URL, or better, add them
+as an associative array to the ``query`` option::
+
+    // it makes an HTTP GET request to https://httpbin.org/get?token=...&name=...
+    $response = $httpClient->request('GET', 'https://httpbin.org/get', [
+        // these values are automatically encoded before including them in the URL
+        'query' => [
+            'token' => '...',
+            'name' => '...',
+        ],
+    ]);
+
+Headers
+~~~~~~~
+
+Use the ``headers`` option to define both the default headers added to all
+requests and the specific headers for each request::
+
+    // this header is added to all requests made by this client
+    $httpClient = HttpClient::create(['headers' => [
+        'Authorization' => 'Bearer ...',
+    ]]);
+
+    // this header is only included in this request
+    $response = $httpClient->request('POST', 'https://...', [
+        'headers' => [
+            'Content-type' => 'text/plain',
+        ],
+    ]);
+
+Uploading Data
+~~~~~~~~~~~~~~
+
+This component provides several methods for uploading data using the ``body``
+option. You can use regular strings, closures and resources and they'll be
+processing automatically when making the requests::
+
+    $response = $httpClient->request('POST', 'https://...', [
+        // defining data using a regular string
+        'body' => 'raw data',
+
+        // using a closure to generate the uploaded data
+        'body' => function () {
+            // ...
+        },
+
+        // using a resource to get the data from it
+        'body' => fopen('/path/to/file', 'r'),
+    ]);
+
+When uploading data with the ``POST`` method, if you don't define the
+``Content-Type`` HTTP header explicitly, Symfony adds the required
+``'Content-Type: application/x-www-form-urlencoded'`` header for you.
+
+Cookies
+~~~~~~~
+
+The HTTP client provided by this component is stateless but handling cookies
+requires a stateful storage (because responses can update cookies and they must
+be used for subsequent requests). That's why this component doesn't handle
+cookies automatically.
+
+You can either handle cookies yourself using the ``Cookie`` HTTP header or use
+the :doc:`BrowserKit component </components/browser_kit>` which provides this
+feature and integrates seamlessly with the HttpClient component.
```

Comment:
Some extra explanation is probably needed here.

I haven't seen anything in symfony/symfony#30413 relating to the BrowserKit component and the documentation for the Browser component states you must implement the `doRequest(Request $request): Response` method, meaning you'd have to:

- Convert `$request` object to `$httpClient->request($method, $url, $options)`
- Convert `Symfony\Contracts\HttpClient\ResponseInterface` to `Symfony\Component\HttpFoundation\Response`

This is non-trivial, and definitely not seamless. Perhaps an idea would be to add a static helper method (like `Symfony\Component\HttpClient\HttpOptions::fromFoundationRequest($request)`) and add an example in this section.

Reply:
Perfectly explained! We need to wait for the merge of the BrowserKit improvements and then we'll add a mention here. Thanks.